tools/buildtools/makeyar.py

```diff
@@ -0,0 +1,178 @@
+# archive begins with a header of non-fixed size, which describes the
+# location of each individual Yaz0 block within the archive itself. This
+# header is followed by each Yaz0 file.
+#
+# The first word (a 4 byte group) of the header indicates the size in bytes of
+# the header itself (also describes the offset of the first Yaz0 block). The
+# rest of the header consists of words describing the offsets of each Yaz0
+# block relative to the end of the header, because the first Yaz0
+# block is omitted from the offsets in the header.
+#
+# Each Yaz0 block is 0xFF-padded to a multiple of 0x10 in size.
+#
+# The entire archive is 0-padded to a multiple of 0x10 in size.
+#
+# The program works by compressing each .data symbol in the input elf file as
+# its own Yaz0 compressed file, appending them in order for the generated
+# archive. Other elf sections are ignored for the resulting yar file.
+#
+# The program also outputs an elf file that's identical to the elf input,
+# but with its .data section zero'ed out completely. This "symbols" elf can be
+# used for referencing each symbol as the whole file were completely
+# uncompressed.
+
+from __future__ import annotations
+
+import argparse
+import dataclasses
+from pathlib import Path
+import struct
+import crunch64
+
+from elftools.elf.elffile import ELFFile
+from elftools.elf.sections import SymbolTableSection
+
+
+def write_word_as_bytes(buff: bytearray, offset: int, word: int):
+    struct.pack_into(f">I", buff, offset, word)
+
+
+@dataclasses.dataclass
+class Symbol:
+    name: str
+    offset: int
+    size: int
+
+
+def get_data_from_elf(elf_path: Path) -> tuple[bytearray, list[Symbol], int]:
+    uncompressed_data = bytearray()
+    symbol_list: list[Symbol] = []
+    data_offset = -1
+
+    with elf_path.open("rb") as elfFile:
+        elf = ELFFile(elfFile)
+        for section in elf.iter_sections():
+            if section.name == ".data":
+                assert len(uncompressed_data) == 0
+                uncompressed_data.extend(section.data())
+                assert len(uncompressed_data) == section["sh_size"]
+                data_offset = section["sh_offset"]
+            elif section.name == ".symtab":
+                assert isinstance(section, SymbolTableSection)
+                for sym in section.iter_symbols():
+                    if sym["st_shndx"] == "SHN_UNDEF":
+                        continue
+                    if sym["st_info"]["type"] != "STT_OBJECT":
+                        continue
+                    symbol_list.append(
+                        Symbol(sym.name, sym["st_value"], sym["st_size"])
+                    )
+    return uncompressed_data, symbol_list, data_offset
+
+
+def align_16(val: int) -> int:
+    return (val + 0xF) & ~0xF
+
+
+def create_archive(
+    uncompressed_data: bytearray, symbol_list: list[Symbol]
+) -> bytearray:
+    archive = bytearray()
+
+    first_entry_offset = (len(symbol_list) + 1) * 4
```

Comment:
```suggestion
    first_entry_offset = (len(symbol_list) + 1) << 2
```
Maybe this isn't the python way to do it, and this is my rust/C/C++ brain thinking about it too much...
But in general a shift would be faster than multiplying

Reply:
Compared to C where a add/shift would translate to one instruction or less, Python has a ton of overhead that makes both options perform about the same

Also in C a good compiler would turn a `*4` into a `<<2` for you anyway!

Reply:
Seems like Python's multiplication operation is faster than its shift operations
https://stackoverflow.com/questions/54047100/why-are-bitwise-operators-slower-than-multiplication-division-modulo
https://stackoverflow.com/questions/37053379/times-two-faster-than-bit-shift-for-python-3-x-integers